subsys/bluetooth/mesh/sensor.c

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <bluetooth/mesh/sensor.h>
+#include <bluetooth/mesh/properties.h>
+#include "sensor.h"
+#include <stdlib.h>
+#include <string.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_MODEL)
+#define LOG_MODULE_NAME bt_mesh_sensor
+#include "common/log.h"
+
+/** Scale away the sensor_value fraction, to allow integer math */
+#define SENSOR_MILL(_val) ((1000000LL * (_val)->val1) + (_val)->val2)
+
+static enum bt_mesh_sensor_cadence
+sensor_cadence(const struct bt_mesh_sensor_threshold *threshold,
+	       const struct sensor_value *curr)
+{
+	s64_t high_mill = SENSOR_MILL(&threshold->range.high);
+	s64_t low_mill = SENSOR_MILL(&threshold->range.low);
+
+	if (high_mill == low_mill) {
+		return BT_MESH_SENSOR_CADENCE_NORMAL;
+	}
+
+	s64_t curr_mill = SENSOR_MILL(curr);
+	bool in_range = (curr_mill >= MIN(low_mill, high_mill) &&
+			 curr_mill <= MAX(low_mill, high_mill));
+
+	return in_range ? threshold->range.cadence : !threshold->range.cadence;
+}
+
+bool bt_mesh_sensor_delta_threshold(const struct bt_mesh_sensor *sensor,
+				    const struct sensor_value *curr)
+{
+	struct sensor_value delta = {
+		curr->val1 - sensor->state.prev.val1,
+		curr->val2 - sensor->state.prev.val2,
+	};
+	s64_t delta_mill = SENSOR_MILL(&delta);
+	s64_t thrsh_mill;
+
+	if (delta_mill < 0) {
+		delta_mill = -delta_mill;
+		thrsh_mill = SENSOR_MILL(&sensor->state.threshold.delta.down);
+	} else {
+		thrsh_mill = SENSOR_MILL(&sensor->state.threshold.delta.up);
+	}
+
+	/* If the threshold value is a perentage, we should calculate the actual
+	 * threshold value relative to the previous value.
+	 */
+	if (sensor->state.threshold.delta.type ==
+	    BT_MESH_SENSOR_DELTA_PERCENT) {
+		s64_t prev_mill = abs(SENSOR_MILL(&sensor->state.prev));
+
+		thrsh_mill = (prev_mill * thrsh_mill) / (100LL * 1000000LL);
+	}
+
+	BT_DBG("Delta: %u (%d - %d) thrsh: %u", (u32_t)(delta_mill / 1000000L),
+	       (s32_t)curr->val1, (s32_t)sensor->state.prev.val1,
+	       (u32_t)(thrsh_mill / 1000000L));
+
+	return (delta_mill > thrsh_mill);
+}
+
+void bt_mesh_sensor_cadence_set(struct bt_mesh_sensor *sensor,
+				enum bt_mesh_sensor_cadence cadence)
+{
+	sensor->state.fast_pub = (cadence == BT_MESH_SENSOR_CADENCE_FAST);
+}
+
+int sensor_status_id_encode(struct net_buf_simple *buf, u8_t len, u16_t id)
+{
+	if ((len > 0 && len <= 16) && id < 2048) {
+		if (net_buf_simple_tailroom(buf) < 2 + len) {
+			return -ENOMEM;
+		}
+
+		net_buf_simple_add_le16(buf, ((len - 1) << 1) | (id << 5));
```

Comment:
Is the order correct here? It seems that format bit should be most significant one, not least. OR the picture in the spec is really misleading. Do you have any other piece of docs or implementation?
![Screen Shot 2020-04-18 at 16 33 32](https://user-images.githubusercontent.com/1160550/79639141-6781f580-8192-11ea-8d6b-7a0d53484729.png)

Reply:
It looks correct to me. Note that the diagram shows the bits from left to right, with the format being bit 0, and the code puts it in as little endian, like the diagram. It is tested against the mesh model test specification, but I don't think that is publicly available.

Reply:
Ok, I will follow your format then in my client app. Thanks for the awesome work on bringing the sensor support to zephyr ecosystem!